src/Middleware/DetectLocaleMiddleware.php

```diff
@@ -0,0 +1,178 @@
+<?php
+
+namespace TractorCow\Fluent\Middleware;
+
+use SilverStripe\Control\Cookie;
+use SilverStripe\Control\HTTPRequest;
+use SilverStripe\Control\Middleware\HTTPMiddleware;
+use SilverStripe\Core\Config\Configurable;
+use TractorCow\Fluent\Extension\FluentDirectorExtension;
+use TractorCow\Fluent\Model\Locale;
+use TractorCow\Fluent\State\FluentState;
+use TractorCow\Fluent\State\LocaleDetector;
+
+/**
+ * DetectLocaleMiddleware will detect if a locale has been requested (or is default) and is not the current
+ * locale, and will redirect the user to that locale if needed.
+ */
+class DetectLocaleMiddleware implements HTTPMiddleware
+{
+    use Configurable;
+
+    /**
+     * IDs to persist the locale in cookies / session in the front end, CMS, etc
+     *
+     * @config
+     * @var string[]
+     */
+     #
+    private static $persist_ids = [
+        'frontend' => 'FluentLocale',
+        'cms' => 'FluentLocale_CMS',
+    ];
+
+    /**
+     * The expiry time in days for a locale persistence cookie
+     *
+     * @config
+     * @var int
+     */
+    private static $persist_cookie_expiry = 90;
+
+    /**
+     * Whether cookies have already been set during {@link setPersistLocale()}
+     *
+     * @var bool
+     */
+    protected $cookiesPersisted = false;
+
+    public function process(HTTPRequest $request, callable $delegate)
+    {
+        $localeDetector = LocaleDetector::singleton()->setRequest($request);
+
+        $state = FluentState::singleton();
+        $state->setLocale($this->getLocale($request));
+
+        $locale = $state->getLocale();
+        $this->setPersistLocale($request, $locale, $this->getPersistKey());
+
+        return $delegate($request);
+    }
+
+    /**
+     * Get the current locale from routing parameters, persistence, browser locale, etc
+     *
+     * @param  HTTPRequest $request
+     * @return string
+     */
+    public function getLocale(HTTPRequest $request)
+    {
+        $state = FluentState::singleton();
+
+        // Check direct request
+        $queryParam = FluentDirectorExtension::config()->get('query_param');
+        $locale = (string) $request->param($queryParam) ?: $request->requestVar($queryParam);
+
+        // Persistant variables
+        if (empty($locale)) {
+            $locale = $this->getPersistLocale($request);
+        }
+
+        // Check browser headers
+        if (empty($locale)) {
+            $locale = LocaleDetector::singleton()->detectBrowserLocale();
+        }
+
+        // Fallback to default if empty or invalid (for this domain)
+        if (empty($locale) || !Locale::getByLocale($locale)) {
+            // If on the frontend, filter locales by the current domain
+            $domain = $state->getIsFrontend($request) ? $state->getDomain() : null;
+
+            /** @var Locale $locale */
+            if ($locale = Locale::getDefault($domain)) {
+                $locale = $locale->Locale;
```

Comment:
Re-assignment of different types to one variable is code smell. Let's separate locale object / strings into different variables.